core/blockchain_impl.go

```diff
@@ -1341,6 +1350,45 @@ func (bc *BlockChainImpl) WriteBlockWithState(
 	return CanonStatTy, nil
 }
 
+// pruneOldTrieAsync prune old trie in async.
+// The task must be completed before the next call, or it will be aborted.
+// Example of why:
+// At State-1, A=1.
+// At State-2, A=0. This means A will be removed from the trie.
+// At State-3, A=1.
+// Assuming that the pruning of (State-1, State-2) is performed after State-3 is flushed to disk, the A will be removed. State-3 will have the wrong value for A.
```

Comment:
Interesting. Do we have a test case for this?

Reply:
yeah, in `TestDiffAndPrune`, It writes dataset1, then dataset2, then dataset1 again.